File: examples/runtime/wasm/runtime.c
#include <stddef.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>

#define MAX_MESSAGE (1024 * 1024)

static char input[MAX_MESSAGE + 8];
static char output[MAX_MESSAGE + 8];
static size_t output_length;

static void append(const char *value) {
  size_t length = strlen(value);
  if (output_length + length > MAX_MESSAGE) _exit(2);
  memcpy(output + output_length, value, length);
  output_length += length;
}

static void append_number(unsigned value) {
  char digits[16];
  size_t length = 0;
  do {
    digits[length++] = (char)('0' + value % 10);
    value /= 10;
  } while (value != 0);
  while (length != 0) {
    char character[2] = {digits[--length], 0};
    append(character);
  }
}

static int field_string_from(const char *source, const char *name, char *target, size_t maximum) {
  char needle[96];
  size_t name_length = strlen(name);
  if (name_length + 5 >= sizeof needle) return 0;
  needle[0] = '"';
  memcpy(needle + 1, name, name_length);
  memcpy(needle + 1 + name_length, "\":\"", 4);
  const char *start = strstr(source, needle);
  if (start == NULL) return 0;
  start += name_length + 4;
  const char *end = start;
  while (*end != 0 && *end != '"') {
    if (*end == '\\') return 0;
    end++;
  }
  size_t length = (size_t)(end - start);
  if (*end != '"' || length + 1 > maximum) return 0;
  memcpy(target, start, length);
  target[length] = 0;
  return 1;
}

static unsigned field_number(const char *source, const char *name) {
  char needle[96];
  size_t name_length = strlen(name);
  needle[0] = '"';
  memcpy(needle + 1, name, name_length);
  memcpy(needle + 1 + name_length, "\":", 3);
  const char *start = strstr(source, needle);
  if (start == NULL) return 0;
  start += name_length + 3;
  unsigned value = 0;
  while (*start >= '0' && *start <= '9') value = value * 10 + (unsigned)(*start++ - '0');
  return value;
}

static void state(const char *source, unsigned *generation, unsigned *calls, unsigned *pending,
                  char *first, size_t first_size) {
  char encoded[2048];
  if (!field_string_from(source, "continuation", encoded, sizeof encoded)) {
    *generation = 0;
    *calls = 0;
    *pending = 0;
    first[0] = 0;
    return;
  }
  const char *cursor = encoded;
  if (*cursor++ != 'g') _exit(2);
  *generation = 0;
  while (*cursor >= '0' && *cursor <= '9') *generation = *generation * 10 + (unsigned)(*cursor++ - '0');
  if (*cursor++ != 'c') _exit(2);
  *calls = 0;
  while (*cursor >= '0' && *cursor <= '9') *calls = *calls * 10 + (unsigned)(*cursor++ - '0');
  if (*cursor++ != 'p') _exit(2);
  *pending = 0;
  while (*cursor >= '0' && *cursor <= '9') *pending = *pending * 10 + (unsigned)(*cursor++ - '0');
  if (*cursor++ != 'f') _exit(2);
  size_t length = strlen(cursor);
  if (length + 1 > first_size) _exit(2);
  memcpy(first, cursor, length + 1);
}

static void continuation(unsigned generation, unsigned calls, unsigned pending, const char *first) {
  append(",\"continuation\":\"g");
  append_number(generation);
  append("c");
  append_number(calls);
  append("p");
  append_number(pending);
  append("f");
  append(first);
  append("\"}");
}

static void prefix(const char *type, const char *id) {
  output_length = 0;
  append("{\"protocol\":\"polici.runtime/v1\",\"type\":\"");
  append(type);
  append("\",\"id\":\"");
  append(id);
}

static void capability(const char *id, unsigned generation, unsigned calls, unsigned sequence,
                       const char *first) {
  prefix("capability-call", id);
  append("\",\"requestId\":\"");
  append(id);
  append("-capability-");
  append_number(sequence);
  append("\",\"sequence\":");
  append_number(sequence);
  append(",\"capability\":\"example:data\",\"operation\":\"read\",\"arguments\":");
  if (sequence == 2) append("{\"page\":{\"tag\":\"integer\",\"value\":\"2\"}}");
  else append("{}");
  continuation(generation, calls, sequence, first);
}

static void result_string(const char *id, const char *value, unsigned generation, unsigned calls) {
  prefix("result", id);
  append("\",\"value\":{\"tag\":\"string\",\"value\":\"");
  append(value);
  append("\"}");
  continuation(generation, calls, 0, "-");
}

static void handle(const char *message) {
  char type[64], id[128], resolver[128];
  if (!field_string_from(message, "type", type, sizeof type) ||
      !field_string_from(message, "id", id, sizeof id)) _exit(2);

  if (strcmp(type, "initialize") == 0) {
    const char *plugin = strstr(message, "\"plugin\":{");
    char name[128], version[128];
    if (plugin == NULL || !field_string_from(plugin, "name", name, sizeof name) ||
        !field_string_from(plugin, "version", version, sizeof version)) _exit(2);
    prefix("initialized", id);
    append("\",\"implementation\":{\"name\":\"");
    append(name);
    append("\",\"version\":\"");
    append(version);
    append("\"},\"capabilities\":[\"example:data\"]");
    continuation(1, 0, 0, "-");
    return;
  }

  unsigned generation, calls, pending;
  char first[1024];
  state(message, &generation, &calls, &pending, first, sizeof first);
  generation++;

  if (strcmp(type, "shutdown") == 0) {
    prefix("stopped", id);
    append("\"}");
    return;
  }

  if (strcmp(type, "call") == 0) {
    if (!field_string_from(message, "resolver", resolver, sizeof resolver)) _exit(2);
    calls++;
    if (strcmp(resolver, "success") == 0) {
      result_string(id, "ok", generation, calls);
      return;
    }
    if (strcmp(resolver, "missing") == 0) {
      prefix("result", id);
      append("\",\"value\":{\"tag\":\"missing\"}");
      continuation(generation, calls, 0, "-");
      return;
    }
    if (strcmp(resolver, "multiple") == 0) {
      capability(id, generation, calls, 1, "-");
      return;
    }
    if (strcmp(resolver, "lifecycle") == 0) {
      prefix("result", id);
      append("\",\"value\":{\"tag\":\"map\",\"entries\":{\"initialized\":{\"tag\":\"integer\",\"value\":\"1\"},\"calls\":{\"tag\":\"integer\",\"value\":\"");
      append_number(calls);
      append("\"}}}");
      continuation(generation, calls, 0, "-");
      return;
    }
    prefix("error", id);
    append("\",\"error\":{\"code\":\"RESOLVER_NOT_FOUND\",\"kind\":\"resolver\",\"message\":\"Unknown resolver\",\"retryable\":false}");
    continuation(generation, calls, 0, "-");
    return;
  }

  if (strcmp(type, "capability-result") == 0) {
    char value[1024];
    const char *result = strstr(message, "\"result\":{");
    if (result == NULL || !field_string_from(result, "value", value, sizeof value)) _exit(2);
    unsigned sequence = field_number(message, "sequence");
    if (pending == 1 && sequence == 1) {
      capability(id, generation, calls, 2, value);
      return;
    }
    if (pending == 2 && sequence == 2) {
      prefix("result", id);
      append("\",\"value\":{\"tag\":\"list\",\"items\":[{\"tag\":\"string\",\"value\":\"");
      append(first);
      append("\"},{\"tag\":\"string\",\"value\":\"");
      append(value);
      append("\"}]}");
      continuation(generation, calls, 0, "-");
      return;
    }
  }
  _exit(2);
}

int main(void) {
  size_t length = 0;
  while (length < MAX_MESSAGE + 4) {
    ssize_t count = read(0, input + length, MAX_MESSAGE + 4 - length);
    if (count < 0) return 2;
    if (count == 0) break;
    length += (size_t)count;
  }
  if (length == 0 || length > MAX_MESSAGE + 4) return 2;
  int framed = input[0] != '{';
  const char *message = input;
  size_t message_length = length;
  if (framed) {
    if (length < 4) return 2;
    message_length = ((unsigned char)input[0] << 24) | ((unsigned char)input[1] << 16) |
                     ((unsigned char)input[2] << 8) | (unsigned char)input[3];
    if (message_length + 4 != length) return 2;
    message = input + 4;
  } else if (input[length - 1] == '\n') {
    message_length--;
  }
  input[(size_t)(message - input) + message_length] = 0;
  handle(message);
  if (framed) {
    unsigned char header[4] = {(unsigned char)(output_length >> 24),
                               (unsigned char)(output_length >> 16),
                               (unsigned char)(output_length >> 8), (unsigned char)output_length};
    if (write(1, header, 4) != 4) return 2;
  }
  if (write(1, output, output_length) != (ssize_t)output_length) return 2;
  if (!framed && write(1, "\n", 1) != 1) return 2;
  return 0;
}

File: examples/runtime/typescript/runtime.ts
// @ts-ignore This bare repository intentionally does not depend on @types/node.
import { readFileSync } from "node:fs";

const PROTOCOL = "polici.runtime/v1";

type Message = Record<string, unknown>;

interface State {
  generation: number;
  calls: number;
  pending: string;
  first: string;
}

const transport = process.argv.includes("--length-prefixed") ? "length-prefixed" : "jsonl";

function main(): void {
  const message = readMessage(readFileSync(0));
  if (message.protocol !== PROTOCOL) throw new Error("Unsupported protocol");
  switch (string(message.type, "type")) {
    case "initialize":
      initialize(message);
      return;
    case "call":
      call(message);
      return;
    case "capability-result":
      capabilityResult(message);
      return;
    case "shutdown":
      shutdown(message);
      return;
    default:
      throw new Error(`Unsupported message ${String(message.type)}`);
  }
}

function initialize(message: Message): void {
  const plugin = record(message.plugin, "plugin");
  const capabilities = array(message.capabilities, "capabilities");
  const activated: string[] = [];
  for (const item of capabilities) {
    const capability = record(item, "capability");
    if (capability.name === "example:data") activated.push("example:data");
  }
  writeMessage({
    protocol: PROTOCOL,
    type: "initialized",
    id: string(message.id, "id"),
    implementation: { name: plugin.name, version: plugin.version },
    capabilities: activated,
    continuation: encodeState({ generation: 1, calls: 0, pending: "", first: "" }),
  });
}

function call(message: Message): void {
  const state = decodeState(string(message.continuation, "continuation"));
  state.generation += 1;
  state.calls += 1;
  const resolver = string(message.resolver, "resolver");
  if (resolver === "success") {
    result(string(message.id, "id"), { tag: "string", value: "ok" }, state);
    return;
  }
  if (resolver === "missing") {
    result(string(message.id, "id"), { tag: "missing" }, state);
    return;
  }
  if (resolver === "lifecycle") {
    result(
      string(message.id, "id"),
      {
        tag: "map",
        entries: {
          initialized: { tag: "integer", value: "1" },
          calls: { tag: "integer", value: String(state.calls) },
        },
      },
      state,
    );
    return;
  }
  if (
    resolver === "capability" ||
    resolver === "multiple" ||
    resolver === "permission-denied" ||
    resolver === "invalid-result"
  ) {
    state.pending = resolver === "multiple" ? "multiple-first" : "single";
    capabilityCall(string(message.id, "id"), 1, "read", message.arguments, state);
    return;
  }
  if (resolver === "capability-timeout") {
    state.pending = "single";
    writeMessage({
      protocol: PROTOCOL,
      type: "capability-call",
      id: string(message.id, "id"),
      requestId: `${string(message.id, "id")}-capability-1`,
      sequence: 1,
      capability: "example:data",
      operation: "read",
      arguments: message.arguments,
      continuation: nextToken(state),
      deadlineUnixMs: 1,
    });
    return;
  }
  if (resolver === "runtime-timeout") {
    let spin = 0;
    while (spin < 1_000_000_000) spin += 1;
    result(string(message.id, "id"), { tag: "string", value: "late" }, state);
    return;
  }
  if (resolver === "large-output") {
    result(string(message.id, "id"), { tag: "string", value: "x".repeat(4096) }, state);
    return;
  }
  if (resolver === "large-log") {
    console.error("x".repeat(4096));
    result(string(message.id, "id"), { tag: "string", value: "logged" }, state);
    return;
  }
  if (resolver === "undeclared") {
    state.pending = "undeclared";
    writeMessage({
      protocol: PROTOCOL,
      type: "capability-call",
      id: string(message.id, "id"),
      requestId: `${string(message.id, "id")}-capability-1`,
      sequence: 1,
      capability: "example:undeclared",
      operation: "read",
      arguments: {},
      continuation: nextToken(state),
    });
    return;
  }
  if (resolver === "invalid-operation") {
    state.pending = "invalid-operation";
    capabilityCall(string(message.id, "id"), 1, "delete", {}, state);
    return;
  }
  runtimeError(
    string(message.id, "id"),
    "RESOLVER_NOT_FOUND",
    "resolver",
    `Unknown resolver ${resolver}`,
    state,
  );
}

function capabilityResult(message: Message): void {
  const state = decodeState(string(message.continuation, "continuation"));
  state.generation += 1;
  const sequence = number(message.sequence, "sequence");
  if (message.error !== undefined) {
    const error = record(message.error, "error");
    runtimeError(
      string(message.id, "id"),
      string(error.code, "error.code"),
      string(error.kind, "error.kind"),
      string(error.message, "error.message"),
      state,
      error.details,
    );
    return;
  }
  const value = record(message.result, "result");
  if (state.pending === "single") {
    state.pending = "";
    result(string(message.id, "id"), value, state);
    return;
  }
  if (state.pending === "multiple-first" && sequence === 1) {
    state.pending = "multiple-second";
    state.first = JSON.stringify(value);
    capabilityCall(
      string(message.id, "id"),
      2,
      "read",
      { page: { tag: "integer", value: "2" } },
      state,
    );
    return;
  }
  if (state.pending === "multiple-second" && sequence === 2) {
    state.pending = "";
    result(
      string(message.id, "id"),
      { tag: "list", items: [JSON.parse(state.first), value] },
      state,
    );
    return;
  }
  runtimeError(
    string(message.id, "id"),
    "INVALID_CONTINUATION",
    "invalid-request",
    "Unexpected capability result",
    state,
  );
}

function shutdown(message: Message): void {
  decodeState(string(message.continuation, "continuation"));
  writeMessage({ protocol: PROTOCOL, type: "stopped", id: string(message.id, "id") });
}

function capabilityCall(
  callId: string,
  sequence: number,
  operation: string,
  arguments_: unknown,
  state: State,
): void {
  writeMessage({
    protocol: PROTOCOL,
    type: "capability-call",
    id: callId,
    requestId: `${callId}-capability-${sequence}`,
    sequence,
    capability: "example:data",
    operation,
    arguments: arguments_,
    continuation: nextToken(state),
  });
}

function result(id: string, value: unknown, state: State): void {
  writeMessage({
    protocol: PROTOCOL,
    type: "result",
    id,
    value,
    continuation: nextToken(state),
  });
}

function runtimeError(
  id: string,
  code: string,
  kind: string,
  message: string,
  state: State,
  details?: unknown,
): void {
  if (details !== undefined) {
    writeMessage({
      protocol: PROTOCOL,
      type: "error",
      id,
      error: { code, kind, message, retryable: false, details },
      continuation: nextToken(state),
    });
    return;
  }
  writeMessage({
    protocol: PROTOCOL,
    type: "error",
    id,
    error: { code, kind, message, retryable: false },
    continuation: nextToken(state),
  });
}

function nextToken(state: State): string {
  return encodeState(state);
}

function encodeState(state: State): string {
  return Buffer.from(
    JSON.stringify({
      generation: state.generation,
      calls: state.calls,
      pending: state.pending,
      first: state.first,
    }),
    "utf8",
  ).toString("base64url");
}

function decodeState(token: string): State {
  const value = JSON.parse(Buffer.from(token, "base64url").toString("utf8")) as unknown;
  const state = record(value, "continuation state");
  return {
    generation: number(state.generation, "state.generation"),
    calls: number(state.calls, "state.calls"),
    pending: string(state.pending, "state.pending"),
    first: string(state.first, "state.first"),
  };
}

function readMessage(input: Uint8Array): Message {
  if (transport === "jsonl") {
    const text = new TextDecoder().decode(input);
    if (!text.endsWith("\n") || text.slice(0, -1).includes("\n"))
      throw new Error("Expected exactly one JSONL message");
    return record(JSON.parse(text), "message");
  }
  if (input.length < 4) throw new Error("Truncated frame header");
  const length = new DataView(input.buffer, input.byteOffset, input.byteLength).getUint32(0, false);
  if (length === 0 || input.length !== length + 4) throw new Error("Invalid frame length");
  return record(JSON.parse(new TextDecoder().decode(input.subarray(4))), "message");
}

function writeMessage(message: unknown): void {
  const encoded = new TextEncoder().encode(JSON.stringify(message));
  if (transport === "jsonl") {
    process.stdout.write(encoded);
    process.stdout.write("\n");
    return;
  }
  const output = new Uint8Array(encoded.length + 4);
  new DataView(output.buffer).setUint32(0, encoded.length, false);
  output.set(encoded, 4);
  process.stdout.write(output);
}

function record(value: unknown, name: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value))
    throw new Error(`${name} must be an object`);
  return value as Record<string, unknown>;
}

function array(value: unknown, name: string): unknown[] {
  if (!Array.isArray(value)) throw new Error(`${name} must be an array`);
  return value;
}

function string(value: unknown, name: string): string {
  if (typeof value !== "string") throw new Error(`${name} must be a string`);
  return value;
}

function number(value: unknown, name: string): number {
  if (typeof value !== "number" || !Number.isSafeInteger(value))
    throw new Error(`${name} must be an integer`);
  return value;
}

main();
